helix-term/src/commands/lsp.rs

```diff
@@ -565,31 +568,87 @@ pub fn goto_reference(cx: &mut Context) {
+}
+
+pub fn signature_help_impl(cx: &mut Context, invoked: SignatureHelpInvoked) {
     let (view, doc) = current!(cx.editor);
     let language_server = language_server!(cx.editor, doc);
     let offset_encoding = language_server.offset_encoding();
 
     let pos = doc.position(view.id, offset_encoding);
 
     let future = language_server.text_document_signature_help(doc.identifier(), pos, None);
+    let was_manually_invoked = invoked == SignatureHelpInvoked::Manual;
 
     cx.callback(
         future,
-        move |_editor, _compositor, response: Option<lsp::SignatureHelp>| {
-            if let Some(signature_help) = response {
-                log::info!("{:?}", signature_help);
-                // signatures
-                // active_signature
-                // active_parameter
-                // render as:
-
-                // signature
-                // ----------
-                // doc
-
-                // with active param highlighted
+        move |editor, compositor, response: Option<lsp::SignatureHelp>| {
+            if !(editor.config().lsp.auto_signature_help
+                || SignatureHelp::visible_popup(compositor).is_some()
+                || was_manually_invoked)
+            {
+                return;
             }
+
+            let response = match response {
+                Some(s) => s,
+                None => {
+                    compositor.remove(SignatureHelp::ID);
+                    return;
+                }
+            };
+            let doc = doc!(editor);
+            let language = doc
+                .language()
+                .and_then(|scope| scope.strip_prefix("source."))
+                .unwrap_or("");
+
+            // If there are no signatures, response will be None, therefore
+            // at least one signature exists.
+            let signature = &response.signatures[response.active_signature.unwrap_or(0) as usize];
+            let mut contents = SignatureHelp::new(
+                signature.label.clone(),
+                language.to_string(),
+                Arc::clone(&editor.syn_loader),
+            );
+
+            let signature_doc = signature.documentation.as_ref().map(|doc| match doc {
+                lsp::Documentation::String(s) => s.clone(),
+                lsp::Documentation::MarkupContent(markup) => markup.value.clone(),
+            });
+            contents.set_signature_doc(signature_doc);
+
+            let active_param_range = || -> Option<(usize, usize)> {
+                let param_idx = signature
+                    .active_parameter
+                    .or(response.active_parameter)
+                    .unwrap_or(0) as usize;
+                let param = signature.parameters.as_ref()?.get(param_idx)?;
+                match &param.label {
+                    lsp::ParameterLabel::Simple(string) => {
+                        let start = signature.label.find(string.as_str())?;
+                        Some((start, start + string.len()))
```

Comment:
I didn't look closely on this. Does this also work with double width characters like CJK? This seemed to be characters length after looking at the code (wasn't mentioned in the comment whether it is character length or terminal width length) and we don't seemed to be checking the width of `required_size`.

Reply:
Syntax highlighting uses char offsets so that is what we need as the final `(start, end)` value. Now, LSP sends the offsets according to the configured `OffsetEncoding` (but we have it [hardcoded to utf8](https://github.com/helix-editor/helix/blob/9a1e5d5887f1da8c2b5fa8caef77f2baff349460/helix-lsp/src/client.rs#L102)) and the same is used by the LSP for `ParameterLabel::LabelOffsets`, so we would have to convert this to char offsets:

```rust
                    lsp::ParameterLabel::LabelOffsets([start, end]) => {
                        Some((*start as usize, *end as usize))
                    }
```

`str.find()` reports the byte index of the start of the pattern, so that would have to be changed too.

Strangely, with a naive test on a bit of CJK and non ASCII text, the parameter highlighting seems to work okay:

![helix-signature-help-cjk](https://user-images.githubusercontent.com/23398472/175787514-5bac103b-f912-49e4-841d-7b85bc01c652.gif)